src/client/inputhandler.h

```diff
@@ -60,44 +72,87 @@ class MyEventReceiver : public IEventReceiver
 #endif
 		// handle mouse events
 		if (event.EventType == irr::EET_MOUSE_INPUT_EVENT) {
-			if (noMenuActive() == false) {
-				left_active = false;
-				middle_active = false;
-				right_active = false;
-			} else {
-				left_active = event.MouseInput.isLeftPressed();
-				middle_active = event.MouseInput.isMiddlePressed();
-				right_active = event.MouseInput.isRightPressed();
-
-				if (event.MouseInput.Event == EMIE_LMOUSE_PRESSED_DOWN) {
-					leftclicked = true;
-				}
-				if (event.MouseInput.Event == EMIE_RMOUSE_PRESSED_DOWN) {
-					rightclicked = true;
-				}
-				if (event.MouseInput.Event == EMIE_LMOUSE_LEFT_UP) {
-					leftreleased = true;
-				}
-				if (event.MouseInput.Event == EMIE_RMOUSE_LEFT_UP) {
-					rightreleased = true;
-				}
+			// handle mouse buttons
+			// left
+			if (event.MouseInput.Event == EMIE_LMOUSE_PRESSED_DOWN) {
+			     	if (keyIsDown["KEY_LBUTTON"]) {
+					keyPressed.set("KEY_LBUTTON");
+			     	}
+				keyDown.set("KEY_LBUTTON");
+				keyIsDown.set("KEY_LBUTTON");
+				keyWasDown.set("KEY_LBUTTON");
+		  	}
+		  	if (event.MouseInput.Event == EMIE_LMOUSE_LEFT_UP) {
+		     		keyDown.unset("KEY_LBUTTON");
+		     		keyIsDown.unset("KEY_LBUTTON");
+	     			keyPressed.unset("KEY_LBUTTON");
+				keyReleased.set("KEY_LBUTTON");	
+		  	}
+
+			// right
+			if (event.MouseInput.Event == EMIE_RMOUSE_PRESSED_DOWN) {
+			     	if (keyIsDown["KEY_RBUTTON"]) {
+					keyPressed.set("KEY_RBUTTON");
+			     	}
+				keyDown.set("KEY_RBUTTON");
+				keyIsDown.set("KEY_RBUTTON");
+				keyWasDown.set("KEY_RBUTTON");
+		  	}
+		  	if (event.MouseInput.Event == EMIE_RMOUSE_LEFT_UP) {
+		     		keyDown.unset("KEY_RBUTTON");
+		     		keyIsDown.unset("KEY_RBUTTON");
+	     			keyPressed.unset("KEY_RBUTTON");
+				keyReleased.set("KEY_RBUTTON");	
+		  	}
+
+			// middle
+			if (event.MouseInput.Event == EMIE_MMOUSE_PRESSED_DOWN) {
+			     	if (keyIsDown["KEY_MBUTTON"]) {
+					keyPressed.set("KEY_MBUTTON");
+			     	}
+				keyDown.set("KEY_MBUTTON");
+				keyIsDown.set("KEY_MBUTTON");
+				keyWasDown.set("KEY_MBUTTON");
+		  	}
+		  	if (event.MouseInput.Event == EMIE_MMOUSE_LEFT_UP) {
+		     		keyDown.unset("KEY_MBUTTON");
+		     		keyIsDown.set("KEY_MBUTTON");
+	     			keyPressed.unset("KEY_MBUTTON");
+				keyReleased.set("KEY_MBUTTON");	
+		  	}
+
+			if (noMenuActive()) {
 				if (event.MouseInput.Event == EMIE_MOUSE_WHEEL) {
 					mouse_wheel += event.MouseInput.Wheel;
 				}
 			}
 		}
+
 		if (event.EventType == irr::EET_LOG_TEXT_EVENT) {
 			dstream << std::string("Irrlicht log: ") + std::string(event.LogEvent.Text)
 			        << std::endl;
 			return true;
 		}
+		if (event.EventType == EET_KEY_INPUT_EVENT) {
+
+		}
```

Comment:
EET_KEY_INPUT_EVENT is handled at the beginning of this method. I will remove it.